python/cudaq/utils/Tensor.py

```diff
@@ -0,0 +1,307 @@
+                A NumPy array representation of the tensor.
+
+        Raises:
+            RuntimeError:
+                If the requested `dtype` doesn't match the `dtype` of the
+                tensor.
+        """
+        if dtype is None:
+            return np.array(self._tensor, copy=False, dtype=dtype)
+        else:
+            if dtype != self._dtype:
+                raise RuntimeError(
+                    f"invalid NumPy `dtype` conversion ({dtype} vs {self._dtype})"
+                )
+            return np.array(self._tensor, copy=False, dtype=dtype)
+
+    def rank(self):
+        """
+        Get the rank (number of dimensions) of the tensor.
+
+        Returns:
+            int: The rank of the tensor.
+        """
+        return self._tensor.rank()
+
+    def size(self):
+        """
+        Get the total number of elements in the tensor.
+
+        Returns:
+            int: The total number of elements.
+        """
+        return self._tensor.size()
+
+    def shape(self):
+        """
+        Get the shape of the tensor.
+
+        Returns:
+            tuple: The shape of the tensor.
+        """
+        return self._tensor.shape()
+
+    def at(self, indices):
+        """
+        Get the value at the specified indices.
+
+        Args:
+            indices (list): List of indices.
+
+        Returns:
+            The value at the specified indices.
+        """
+        return self._tensor.at(indices)
+
+    def copy(self, data):
+        """
+        Copy data into the tensor.
+
+        Args:
+            data: The array-like data to copy into the tensor.
+        """
+        np_data = np.array(data, dtype=self._scalar_type_to_numpy(self._dtype))
+        self._tensor.copy(np_data)
+
+    def take(self, data):
+        """
+        Take data from the tensor.
+
+        Args:
+            data: The array-like data to take from the tensor.
+        """
+        np_data = np.array(data,
+                           copy=False,
+                           dtype=self._scalar_type_to_numpy(self._dtype))
+        self._tensor.take(np_data)
+
+    def borrow(self, data):
+        """
+        Borrow data from the tensor.
+
+        Args:
+            data: The array-like data to borrow from the tensor.
+        """
+        np_data = np.array(data,
+                           copy=False,
+                           dtype=self._scalar_type_to_numpy(self._dtype))
+        self._tensor.borrow(np_data)
+
+    def dump(self):
+        """
+        Dump the contents of the tensor.
+        """
+        self._tensor.dump()
+
+    def __str__(self):
+        """
+        Print a NumPy-like string representation of the tensor 
+        """
+        return str(np.array(self._tensor, copy=False))
+
+    @property
+    def dtype(self):
+        """
+        Get the data type of the tensor.
+
+        Returns:
+            `numpy.dtype`: The data type of the tensor elements.
+        """
+        return self._dtype
```

Comment:
we already have a `shape()` method, would you like me to replace it with a property?

Reply:
The shape information is buried in the C++ object, so it would seem we need a function, no?

Reply:
Was just thinking it would be convenient to have both 
```python
@property 
def shape(): return self._tensor.shape()
```
maybe its not possible to have both.